dm_control/locomotion/walkers/assets/dog_v2/add_muscles.py

```diff
@@ -0,0 +1,435 @@
+    else:
+      if body.name in ["upper_leg_L",
+                       "upper_leg_R", "lower_leg_L", "lower_leg_R",
+                       "foot_L", "foot_R", "toe_L", "toe_R",
+                       "lower_arm_L", "lower_arm_R",
+                       "hand_L", "hand_R", "finger_L", "finger_R"]:
+        continue
+
+    if dist_new < dist:
+      dist = dist_new
+      closest_geom_body = body
+
+  return closest_geom_body
+
+
+def calculate_transformation(element):
+  """
+    Calculates the transformation matrix from the root to the given element.
+
+    This function iterates through the parent elements of the given element and
+    calculates the transformation matrices from the root until reaching the
+    given element. It considers the position and orientation (quaternion)
+    attributes of each element to construct the transformation matrices.
+    The final transformation matrix represents the cumulative transformation
+    from the root to the given element.
+
+    Args:
+        element: The MuJoCo element for which the transformation matrix
+          is calculated.
+
+    Returns:
+        T: The transformation matrix from the root to the given element.
+  """
+  # Calculate all transformation matrices from root until this element
+  all_transformations = []
+
+  while type(element) == mjcf_module.element._ElementImpl:
+    if element.pos is not None:
+      pos = np.array(element.pos, dtype=float)
+    else:
+      pos = np.zeros(3)
+
+    if element.quat is not None:
+      rot = Rot.from_quat(element.quat).as_matrix()
+    else:
+      rot = Rot.identity().as_matrix()
+
+    all_transformations.append(
+      np.vstack(
+        (np.hstack((rot, pos.reshape([-1, 1]))), np.array([0, 0, 0, 1])))
+    )
+
+    element = element.parent
+
+  # Apply all transformations
+  T = np.eye(4)
+  for transformation in reversed(all_transformations):
+    T = np.matmul(T, transformation)
+
+  return T
+
+
+def add_muscles(model, scale_multiplier, muscle_dynamics,
+                asset_dir, lengthrange_from_joints):
+  physics = mjcf_module.Physics.from_mjcf_model(model)
+
+  mjcf = model
+
+  muscle_meshes_path = asset_dir + '/muscles/'
+  bones_path = asset_dir
+
+  bones = [f for f in listdir(bones_path)
+           if isfile(join(bones_path, f)) and f[0] != '.']
+
+  bones_kd_trees = {}
+  bones_meshes = {}
+  for bone in bones:
+    if "BONE" in bone and "simpl" not in bone and "Lingual" not in bone and "cartilage" not in bone:
+      bone_mesh = trimesh.load_mesh(bones_path + '/' + bone, process=False)
+      bones_kd_trees[bone[4:-4]] = KDTree(bone_mesh.vertices)
+      bones_meshes[bone[4:-4]] = bone_mesh
+
+  muscle = mjcf.default.default["muscle"]
+
+  flexors = muscle.add("default", dclass="flexors")
+  flexors.tendon.rgba = [0.5, 0, 0, 1]
+
+  extensors = muscle.add("default", dclass="extensors")
+  extensors.tendon.rgba = [0.5, 0, 0, 1]
+
+  length_range = mjcf.compiler.lengthrange
+  length_range.inttotal = 50
```

Comment:
If it's plausible that any other code could depend on these consider making them file level constants rather than hard coding them here.